In [26]:
import pandas as pd
import numpy as np

Задание 1

In [21]:
horses_stat = pd.read_csv('horse_data.csv', header=None, na_values='?')
horses_stat.head(5)


0   1        2     3      4     5    6    7    8    9   ...    18    19  \
0  2.0   1   530101  38.5   66.0  28.0  3.0  3.0  NaN  2.0  ...  45.0   8.4   
1  1.0   1   534817  39.2   88.0  20.0  NaN  NaN  4.0  1.0  ...  50.0  85.0   
2  2.0   1   530334  38.3   40.0  24.0  1.0  1.0  3.0  1.0  ...  33.0   6.7   
3  1.0   9  5290409  39.1  164.0  84.0  4.0  1.0  6.0  2.0  ...  48.0   7.2   
4  2.0   1   530255  37.3  104.0  35.0  NaN  NaN  6.0  2.0  ...  74.0   7.4   

    20   21   22  23     24  25  26  27  
0  NaN  NaN  2.0   2  11300   0   0   2  
1  2.0  2.0  3.0   2   2208   0   0   2  
2  NaN  NaN  1.0   2      0   0   0   1  
3  3.0  5.3  2.0   1   2208   0   0   1  
4  NaN  NaN  2.0   2   4300   0   0   2  

[5 rows x 28 columns]

In [68]:
horses_stat.columns = ['surgery', 
                       'Age', 
                       'Hospital_Number', 
                       'rectal_temperature', 
                       'pulse', 
                       'respiratory_rate', 
                       'temperature_of_extremities',
                       'peripheral_pulse',
                       'mucous_membranes',
                       'capillary_refill_time',
                       'pain',
                       'peristalsis',
                       'abdominal_distension',
                       'nasogastric_tube', 
                       'nasogastric_reflux', 
                       'nasogastric_reflux_PH', 
                       'rectal_examination', 
                       'abdomen', 
                       'packed_cell_volume', 
                       'total_protein',
                       'abdominocentesis_appearance',
                       'abdomcentesis_total_protein',
                       'outcome',
                       'surgical_lesion',
                       'type_of_lesion',
                       'type_of_lesion',
                       'type_of_lesion',
                       'cp_data'
                      ]
horses_stat.head(10)

surgery  Age  Hospital_Number  rectal_temperature  pulse  respiratory_rate  \
0      2.0    1           530101                38.5   66.0              28.0   
1      1.0    1           534817                39.2   88.0              20.0   
2      2.0    1           530334                38.3   40.0              24.0   
3      1.0    9          5290409                39.1  164.0              84.0   
4      2.0    1           530255                37.3  104.0              35.0   
5      2.0    1           528355                 NaN    NaN               NaN   
6      1.0    1           526802                37.9   48.0              16.0   
7      1.0    1           529607                 NaN   60.0               NaN   
8      2.0    1           530051                 NaN   80.0              36.0   
9      2.0    9          5299629                38.3   90.0               NaN   

   temperature_of_extremities  peripheral_pulse  mucous_membranes  \
0                         3.0               3.0               NaN   
1                         NaN               NaN               4.0   
2                         1.0               1.0               3.0   
3                         4.0               1.0               6.0   
4                         NaN               NaN               6.0   
5                         2.0               1.0               3.0   
6                         1.0               1.0               1.0   
7                         3.0               NaN               NaN   
8                         3.0               4.0               3.0   
9                         1.0               NaN               1.0   

   capillary_refill_time  ...  packed_cell_volume  total_protein  \
0                    2.0  ...                45.0            8.4   
1                    1.0  ...                50.0           85.0   
2                    1.0  ...                33.0            6.7   
3                    2.0  ...                48.0            7.2   
4                    2.0  ...                74.0            7.4   
5                    1.0  ...                 NaN            NaN   
6                    1.0  ...                37.0            7.0   
7                    1.0  ...                44.0            8.3   
8                    1.0  ...                38.0            6.2   
9                    1.0  ...                40.0            6.2   

   abdominocentesis_appearance  abdomcentesis_total_protein  outcome  \
0                          NaN                          NaN      2.0   
1                          2.0                          2.0      3.0   
2                          NaN                          NaN      1.0   
3                          3.0                          5.3      2.0   
4                          NaN                          NaN      2.0   
5                          NaN                          NaN      1.0   
6                          NaN                          NaN      1.0   
7                          NaN                          NaN      2.0   
8                          NaN                          NaN      3.0   
9                          1.0                          2.2      1.0   

   surgical_lesion  type_of_lesion  type_of_lesion  type_of_lesion  cp_data  
0                2           11300               0               0        2  
1                2            2208               0               0        2  
2                2               0               0               0        1  
3                1            2208               0               0        1  
4                2            4300               0               0        2  
5                2               0               0               0        2  
6                1            3124               0               0        2  
7                1            2208               0               0        2  
8                1            3205               0               0        2  
9                2           

In [69]:
h = horses_stat[['Age', 
             'pulse', 
             'capillary_refill_time',
             'total_protein', 
             'outcome', 
             'pain',
             'abdominal_distension', 
             'abdomen']]
h.head()

Age  pulse  capillary_refill_time  total_protein  outcome  pain  \
0    1   66.0                    2.0            8.4      2.0   5.0   
1    1   88.0                    1.0           85.0      3.0   3.0   
2    1   40.0                    1.0            6.7      1.0   3.0   
3    9  164.0                    2.0            7.2      2.0   2.0   
4    1  104.0                    2.0            7.4      2.0   NaN   

   abdominal_distension  abdomen  
0                   4.0      5.0  
1                   2.0      2.0  
2                   1.0      1.0  
3                   4.0      NaN  
4                   NaN      NaN

In [70]:
print(np.max(h))

Age                        9.0
pulse                    184.0
capillary_refill_time      3.0
total_protein             89.0
outcome                    3.0
pain                       5.0
abdominal_distension       4.0
abdomen                    5.0
dtype: float64


Из данных по максимальным значениям:

Age проверяем на выбросы

pulse в рамках (у лошадей пульс может быть до 200)

capillary_refill_time  проверяем (должны быть значения только 1 и 2) 

total_protein в рамках

outcome в рамках

pain в рамках

abdominal_distension в рамках

abdomen в рамках


In [71]:
print(np.min(h))

Age                       1.0
pulse                    30.0
capillary_refill_time     1.0
total_protein             3.3
outcome                   1.0
pain                      1.0
abdominal_distension      1.0
abdomen                   1.0
dtype: float64


In [72]:
range_ = h.max() - h.min()
print(range_)

Age                        8.0
pulse                    154.0
capillary_refill_time      2.0
total_protein             85.7
outcome                    2.0
pain                       4.0
abdominal_distension       3.0
abdomen                    4.0
dtype: float64


In [73]:
print(np.mean(h))

Age                       1.640000
pulse                    71.913043
capillary_refill_time     1.305970
total_protein            24.456929
outcome                   1.551839
pain                      2.951020
abdominal_distension      2.266393
abdomen                   3.692308
dtype: float64


In [74]:
print(h.mode())

   Age  pulse  capillary_refill_time  total_protein  outcome  pain  \
0  1.0   48.0                    1.0            6.5      1.0   3.0   
1  NaN    NaN                    NaN            7.0      NaN   NaN   

   abdominal_distension  abdomen  
0                   1.0      5.0  
1                   NaN      NaN  


In [75]:
print(h.median())

Age                       1.0
pulse                    64.0
capillary_refill_time     1.0
total_protein             7.5
outcome                   1.0
pain                      3.0
abdominal_distension      2.0
abdomen                   4.0
dtype: float64


Задание 2

Проверяем Age на выбросы

In [91]:
q1 = h.Age.quantile(0.25)
q3 = h.Age.quantile(0.75)
iqr = q3 - q1
lower_bound = q1 - (1.5 * iqr) 
upper_bound = q3 + (1.5 * iqr)
remove_outliers = h[h.Age.between(lower_bound, upper_bound, inclusive=True)]
remove_outliers['Age']

0      1
1      1
2      1
4      1
5      1
      ..
295    1
296    1
297    1
298    1
299    1
Name: Age, Length: 276, dtype: int64

In [92]:
pd.concat([h, remove_outliers]).drop_duplicates(keep=False)['Age']

3      9
9      9
13     9
16     9
23     9
39     9
41     9
55     9
74     9
75     9
82     9
103    9
109    9
135    9
191    9
212    9
227    9
229    9
244    9
252    9
255    9
262    9
275    9
287    9
Name: Age, dtype: int64

Age

1 = Adult horse
2 = Young (< 6 months)

Выбросы =9. 

Скорее всего ошибка ввода. По количеству 23 шт (в соотношении с 299). Отношу их к "2"

Проверяем capillary_refill_time на значение 3 (данные из max). По условиям там только 1 и 2

In [108]:
h[h.capillary_refill_time == 3]

Age  pulse  capillary_refill_time  total_protein  outcome  pain  \
78    1    NaN                    3.0            5.9      2.0   5.0   
85    1   54.0                    3.0            6.2      1.0   3.0   

    abdominal_distension  abdomen  
78                   3.0      5.0  
85                   2.0      4.0

Значений всего 2 шт (по отношению к 299), поэтому исправляем на 2.

Задание 3

In [110]:
for col in h.columns:
    pct_missing = h[col].isnull().mean()
    print(f'{col} - {pct_missing :.1%}')

Age - 0.0%
pulse - 8.0%
capillary_refill_time - 10.7%
total_protein - 11.0%
outcome - 0.3%
pain - 18.3%
abdominal_distension - 18.7%
abdomen - 39.3%


Работаем только total_protein, pain, abdominal_distension, abdomen, так как для остальных показателей оставшихся данных достаточно для анализа

In [155]:
h.total_protein.value_counts()

7.5     46
7.0     15
6.5     15
6.6     11
7.2      9
        ..
89.0     1
59.0     1
53.0     1
46.0     1
85.0     1
Name: total_protein, Length: 81, dtype: int64

In [156]:
fill_median = horses_stat

fill_median.total_protein.fillna(horses_stat.total_protein.median(), inplace=True)
fill_median.total_protein.fillna(horses_stat.total_protein.median()).isna().sum()

0

In [157]:
h.pain.value_counts()

3.0    67
2.0    59
5.0    42
4.0    39
1.0    38
Name: pain, dtype: int64

In [158]:
fill_mode = horses_stat
fill_mode.pain.fillna(horses_stat.pain.mode()[0], inplace=True)

In [159]:
h.abdominal_distension.value_counts()

1.0    76
3.0    65
2.0    65
4.0    38
Name: abdominal_distension, dtype: int64

In [160]:
fill_mode_1 = horses_stat
fill_mode_1.abdominal_distension.fillna(horses_stat.abdominal_distension.mode()[0], inplace=True)

In [161]:
h.abdomen.value_counts()

5.0    79
4.0    43
1.0    28
2.0    19
3.0    13
Name: abdomen, dtype: int64

In [162]:
horses_stat['abdomen'].fillna(horses_stat.abdomen.mode()[0], inplace=True)